### PR TITLE
feat(agg-index): generate agg indexes after data inserted

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

* Add a new create sync aggregating index syntax.
* Synchronize generate aggregating indexes after insert into.
* Add `sync_creation` to `IndexMeta` to indicate if index is sync created.

copy/replace into and compact will be done in future PRs.

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12464)

### CODE REVIEW

File: src/query/service/src/interpreters/interpreter_insert.rs [hunk 1]
```diff
@@ -262,6 +264,20 @@ impl Interpreter for InsertInterpreter {
             append_mode,
         )?;
 
+        let refresh_agg_index_desc = RefreshAggIndexDesc {
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/interpreter_insert.rs:267]: Insert is not the main path for data ingest to Databend, COPY/REPLACE is.
Why we need to refresh the index after insert?

File: src/query/service/src/interpreters/common/refresh_aggregating_index.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+) -> Result<()> {
+    if pipeline.is_empty() {
+        return Ok(());
+    }
+
+    if ctx
+        .get_settings()
+        .get_enable_sync_refresh_aggregating_index()?
+    {
+        pipeline.set_on_finished(move |err| {
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/common/refresh_aggregating_index.rs:59]: This maybe a issue if the pipeline already have `set_on_finished`.

File: src/query/service/src/interpreters/interpreter_insert.rs (hunk 1 above)
Comment (BohuTANG) [reply src/query/service/src/interpreters/interpreter_insert.rs:267]: Oh, there is a `get_enable_sync_refresh_aggregating_index()` settings already.
The name `enable_refresh_aggregating_index_after_write` is better?

File: src/query/service/src/interpreters/common/refresh_aggregating_index.rs (hunk 2 above)
Comment (BohuTANG) [reply src/query/service/src/interpreters/common/refresh_aggregating_index.rs:59]:
```
   pub fn set_on_finished<F: FnOnce(&Option<ErrorCode>) -> Result<()> + Send + Sync + 'static>(
        &mut self,
        f: F,
    ) {
        if let Some(on_finished) = self.on_finished.take() {
            self.on_finished = Some(Box::new(move |may_error| {
                on_finished(may_error)?;
                f(may_error)
            }));

            return;
        }

        self.on_finished = Some(Box::new(f));
    }
```

The `set_on_finished` may should return error if we have already `set_on_finished`, instead use the first set one.  cc @zhang2014

File: src/query/service/src/interpreters/interpreter_insert.rs (hunk 1 above)
Comment (ariesdevil) [reply src/query/service/src/interpreters/interpreter_insert.rs:267]: > Insert is not the main path for data ingest to Databend, COPY/REPLACE is.
Yep, I'll go on other ways after the insert is completed.

> Why we need to refresh the index after insert? The agg index refresh will run in the background service not the query.
Users may want the index generation in real time. cc @sundy-li

File: src/query/service/src/interpreters/common/refresh_aggregating_index.rs (hunk 2 above)
Comment (ariesdevil) [reply src/query/service/src/interpreters/common/refresh_aggregating_index.rs:59]: Exactly this is a potential issue, the current pipeline implementation does not support multiple `set_on_finished` or something like `finish chain`.
Currently, we need manually check if the worked-on pipeline has a `set_on_finished`, if so, we need to modify the code in the existing `set_on_finished`.

File: src/query/service/src/interpreters/interpreter_insert.rs (hunk 1 above)
Comment (sundy-li) [reply src/query/service/src/interpreters/interpreter_insert.rs:267]: This should not be a setting but another kind of aggregating index.

```
CREATE sync | [async]  AGGREGATING INDEX ....
```

Comment (sundy-li) [reply src/query/service/src/interpreters/interpreter_insert.rs:267]: >  Users may want the index generation in real time.

That's right, we should provide this option to let users choose the suitable mode.

File: src/query/service/src/interpreters/common/refresh_aggregating_index.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            index_meta,
+            None,
+            Some(segment_locs),
+        )
+        .await
+}
+
+async fn refresh_agg_index(ctx: Arc<QueryContext>, desc: RefreshAggIndexDesc) -> Result<()> {
+    let plans = generate_refresh_index_plan(ctx.clone(), desc).await?;
+    let refresh_agg_index_interpreter =
```
Comment (ariesdevil) [inline src/query/service/src/interpreters/common/refresh_aggregating_index.rs:138]: Should be running parallel for multiple indexes.

File: src/query/service/src/interpreters/common/refresh_aggregating_index.rs (hunk 2 above)
Comment (zhang2014) [reply src/query/service/src/interpreters/common/refresh_aggregating_index.rs:59]: set_on_finished is chain.

**Review by drmingdrmer — CHANGES_REQUESTED** [review]
> Reviewed 2 of 24 files at r1, 9 of 34 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12464)* status: 11 of 45 files reviewed, 5 unresolved discussions (waiting on @ariesdevil, @BohuTANG, @RinChanNOWWW, and @sundy-li)

___
*[`src/meta/api/src/schema_api_impl.rs` line 1238 at r3](https://reviewable.io/reviews/datafuselabs/databend/12464#-Nbx6u-17MU2cokv_89W:-Nbx6u-17MU2cokv_89X:b80nwby) ([raw file](https://github.com/datafuselabs/databend/blob/0f4419c16db2635994f0243d9eb098060930814e/src/meta/api/src/schema_api_impl.rs#L1238)):*
> ```XML
>     #[logcall::logcall("debug")]
>     #[minitrace::trace]
>     async fn list_indexes_by_table_id(
> ```

Do not forget to add the corresponding test to `schema_api_test_suite.rs`.
___
*[`src/meta/app/src/schema/index.rs` line 107 at r3](https://reviewable.io/reviews/datafuselabs/databend/12464#-Nbx62hf0hY_PgO5zJUD:-Nbx62hf0hY_PgO5zJUE:byrv1aj) ([raw file](https://github.com/datafuselabs/databend/blob/0f4419c16db2635994f0243d9eb098060930814e/src/meta/app/src/schema/index.rs#L107)):*
> ```XML
>     // if true, index will create after data written to databend,
>     // no need execute refresh index manually.
>     pub sync_creation: bool,
> ```

`sync_creation` is a bit obscure if it is meant to **update-index**.

**Comment by ariesdevil** [comment]
> `sync_creation` is a bit obscure if it is meant to update-index.

`sync_creation` means the index generated after data is written, no need manually execute refresh index. Actually, I'm not sure what this field should be named, `sync_creation` is the answer ChatGPT given.

**Comment by drmingdrmer** [comment]
> > `sync_creation` is a bit obscure if it is meant to update-index.
> 
> `sync_creation` means the index generated after data is written, no need manually execute refresh index. Actually, I'm not sure what this field should be named, `sync_creation` is the answer ChatGPT given.

Whatever. There may be or may not be a better name, and it should not be a blocking issue.

File: src/query/service/tests/it/storages/testdata/settings_table.txt [hunk 4]
```diff
@@ -15,6 +15,7 @@ DB.Table: 'system'.'settings', Table: settings-table_id:1, ver:0, Engine: System
 | 'enable_hive_parquet_predict_pushdown'         | '1'            | '1'            | 'SESSION' | 'Enable hive parquet predict pushdown  by setting this variable to 1, default value: 1'                                                                                               | 'UInt64' |
 | 'enable_query_result_cache'                    | '0'            | '0'            | 'SESSION' | 'Enables caching query results to improve performance for identical queries.'                                                                                                         | 'UInt64' |
 | 'enable_recluster_after_write'                 | '1'            | '1'            | 'SESSION' | 'Enables re-clustering after write(copy/replace-into).'                                                                                                                               | 'UInt64' |
+| 'enable_refresh_aggregating_index_after_write' | '0'            | '0'            | 'SESSION' | 'Refresh aggregating index after new data written'                                                                                                                                    | 'UInt64' |
```
Comment (RinChanNOWWW) [inline src/query/service/tests/it/storages/testdata/settings_table.txt:18]: If the user create a sync agg index, but disable this setting, and insert data. Agg index blocks belonging to this data will never be created?

Comment (ariesdevil) [reply src/query/service/tests/it/storages/testdata/settings_table.txt:18]: Yep, this setting is just for now and will be enabled default when all parts(copy/replace into) are finished.

File: src/query/service/src/sessions/query_ctx.rs [hunk 5]
```diff
@@ -126,6 +120,7 @@ impl QueryContext {
             shared,
             query_settings,
             fragment_id: Arc::new(AtomicUsize::new(0)),
+            segment_locations: Arc::new(RwLock::new(Vec::new())),
```
Comment (RinChanNOWWW) [inline src/query/service/src/sessions/query_ctx.rs:123]: I think we can change a better name to indicate these `segment_locations` are generated by insertion. And add some comments.

**Review by drmingdrmer — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12464)* status: 12 of 45 files reviewed, 7 unresolved discussions (waiting on @ariesdevil, @BohuTANG, @RinChanNOWWW, and @sundy-li)

**Review by drmingdrmer — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12464)* status: 12 of 45 files reviewed, 7 unresolved discussions (waiting on @ariesdevil, @BohuTANG, @RinChanNOWWW, and @sundy-li)

**Review by drmingdrmer — APPROVED** [review]
> Reviewed 3 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12464)* status: 12 of 45 files reviewed, 7 unresolved discussions (waiting on @ariesdevil, @BohuTANG, @RinChanNOWWW, and @sundy-li)

**Review by RinChanNOWWW — APPROVED** [review]

**Pull request merged**